File: Projects/SINAB_Entidades/Helpers/EstablecimientoHelpers/Distribucion.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Design;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Security.Cryptography;
using System.Transactions;
using System.Web.UI.WebControls;

namespace SINAB_Entidades.Helpers.EstablecimientoHelpers
{
    public static class Distribucion
    {
        /// <summary>
        /// Obtiene distribución por establecimiento e id conocido
        /// </summary>
        /// <param name="idDistribucion"></param>
        /// <param name="idEstablecimiento"></param>
        /// <returns>la distribución que concuerda con los parametros pedidos</returns>
        public static SAB_EST_DISTRIBUCION Obtener(int idDistribucion, int  idEstablecimiento)
        {
            using (var db = new SinabEntities())
            {
                return
                    db.SAB_EST_DISTRIBUCION.
                    Include(d => d.SAB_CAT_ALMACENES).
                    Include(d => d.SAB_CAT_SUMINISTROS).
                    FirstOrDefault( d => 
                        d.IDDISTRIBUCION == idDistribucion && 
                        d.IDESTABLECIMIENTO == idEstablecimiento);
            }
        }

        public static SAB_EST_DISTRIBUCION Obtener(SAB_EST_DISTRIBUCION distribucion)
        {
            using (var db = new SinabEntities())
            {
                return Obtener(db, distribucion);
            }
        }

        public static SAB_EST_DISTRIBUCION Obtener(SinabEntities db, SAB_EST_DISTRIBUCION distribucion)
        {
                return
                    db.SAB_EST_DISTRIBUCION.
                    Include(d => d.SAB_CAT_ALMACENES).
                    Include(d => d.SAB_CAT_SUMINISTROS).
                    FirstOrDefault(d =>
                        d.IDDISTRIBUCION == distribucion.IDDISTRIBUCION &&
                        d.IDESTABLECIMIENTO == distribucion.IDESTABLECIMIENTO);
        }

        public static SAB_EST_DISTRIBUCION ObtenerCerrados(int idDistribucion, int idEstablecimientoDistribucion)
        {
            using (var db = new SinabEntities())
            {
                return (from d in db.SAB_EST_DISTRIBUCION

                        join de in db.SAB_EST_DISTRIBUCIONESTABLECIMIENTO on new { d.IDDISTRIBUCION, d.IDESTABLECIMIENTO } equals
                        new { de.IDDISTRIBUCION, de.IDESTABLECIMIENTO }

                        join dc in db.SAB_EST_DISTRIBUCIONESCERRADAS on new { d.IDDISTRIBUCION, d.IDESTABLECIMIENTO } equals
                        new { IDDISTRIBUCION = (Int32)dc.IDDISTRIBUCION, IDESTABLECIMIENTO = (Int32)dc.IDESTABLECIMIENTO }

                        where (d.ESTADO == 2 || d.ESTADO == 3) && d.IDDISTRIBUCION == idDistribucion && de.IDESTABLECIMIENTODISTRIBUCION == idEstablecimientoDistribucion
                        select d).FirstOrDefault();
            }
        }

        /// <summary>
        /// Recupera el mayor id actual de distribuciones y le suma 1 para calcular el nuevo maximo
        /// </summary>i
        /// <returns>retorna el maximo id calculado en ese momento</returns>
        public static long ObtenerMaxId(int idEstablecimiento)
        {
            using (var db = new SinabEntities())
            {
                return ObtenerMaxId(db, idEstablecimiento);
            }
        }
        public static int ObtenerMaxId(SinabEntities db, int idEstablecimeinto)
        {
                var rec = db.SAB_EST_DISTRIBUCION.Where(d => d.IDESTABLECIMIENTO == idEstablecimeinto).Max(l => l.IDDISTRIBUCION);
                return rec + 1;
        }

       
        /// <summary>
        /// Modifica una distribución existente
        /// </summary>
        /// <param name="distribucion">objeto con el que actualizará la distribución</param>
        /// <param name="db">contexto</param>
        public static void Modificar(SAB_EST_DISTRIBUCION distribucion, SinabEntities db)
        {
            var d = Obtener(db, distribucion);
            if (d == null) throw (new Exception("Error, no se encontro la distribución a actualizar"));
            d.IDALMACEN = distribucion.IDALMACEN;
            d.DESCRIPCION = distribucion.DESCRIPCION;
            d.FECHACORTE = distribucion.FECHACORTE;
            d.MESESCPM = distribucion.MESESCPM;
            d.MESESADMINISTRACION = distribucion.MESESADMINISTRACION;
            d.MESESSEGURIDAD = distribucion.MESESSEGURIDAD;
            d.MESESDISTRIBUCION = distribucion.MESESDISTRIBUCION;
            d.AUUSUARIOMODIFICACION = distribucion.AUUSUARIOMODIFICACION;
            d.AUFECHAMODIFICACION = distribucion.AUFECHAMODIFICACION;
            db.SaveChanges();
        }
        public static void Modificar(SAB_EST_DISTRIBUCION distribucion)
        {
            using (var db = new SinabEntities())
            {
                Modificar(distribucion, db);
            }
        }

        /// <summary>
        /// Crea una nueva distribucion
        /// </summary>
        /// <param name="distribucion">Objeto a crear</param>
        /// <param name="listado">listado de indice de productos</param>
        public static void Crear(SAB_EST_DISTRIBUCION distribucion, List<int> listado = null)
        {
            using (var ts = new TransactionScope())
            {
                try
                {
                    using (var db = new SinabEntities())
                    {
                        distribucion.IDDISTRIBUCION = ObtenerMaxId(db, distribucion.IDESTABLECIMIENTO);
                        db.SAB_EST_DISTRIBUCION.AddObject(distribucion);
                        db.SaveChanges();

                        DistribucionProductos.AgregarVarios(db, distribucion, listado);
                        DistribucionProductosLote.AgregarVarios(db, distribucion, listado);
                        DistribucionEstablecimiento.Agregar(db, distribucion);
                    }
                    ts.Complete();
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }

        /// <summary>
        /// Obtiene todas las distribuciones del usuario parametrizado
        /// </summary>
        /// <param name="idEstablecimiento"></param>
        /// <param name="usr"></param>
        /// <returns>Lista de distribuciones</returns>
        public static List<SAB_EST_DISTRIBUCION> ObtenerTodos(int idEstablecimiento, string usr)
        {
            using (var db = new SinabEntities())
            {
                return (from d in db.SAB_EST_DISTRIBUCION
                            .Include(b=> b.SAB_CAT_SUMINISTROS)
                            .Include(b=> b.SAB_CAT_ALMACENES)
                    where
                        d.IDESTABLECIMIENTO == idEstablecimiento &&
                        (d.AUUSUARIOCREACION == usr || d.AUUSUARIOMODIFICACION == usr)
                        select d).OrderByDescending(d => d.IDDISTRIBUCION).ToList();
            }
        }


        public static List<SAB_EST_DISTRIBUCION> ObtenerTodosCerrados(int idEstablecimientoDistribucion, int anio)
        {
            using (var db = new SinabEntities())
            {
                return (from d in db.SAB_EST_DISTRIBUCION

                    join de in db.SAB_EST_DISTRIBUCIONESTABLECIMIENTO on new {d.IDDISTRIBUCION, d.IDESTABLECIMIENTO}
                        equals
                        new {de.IDDISTRIBUCION, de.IDESTABLECIMIENTO}

                    join dc in db.SAB_EST_DISTRIBUCIONESCERRADAS on new {d.IDDISTRIBUCION, d.IDESTABLECIMIENTO} equals
                        new
                        {
                            IDDISTRIBUCION = (Int32) dc.IDDISTRIBUCION,
                            IDESTABLECIMIENTO = (Int32) dc.IDESTABLECIMIENTO
                        }
                    where (d.ESTADO == 2 || d.ESTADO == 3) &&
                          de.IDESTABLECIMIENTODISTRIBUCION == idEstablecimientoDistribucion &&
                          d.FECHADISTRIBUCION.Year == anio
                    select d).Distinct().OrderByDescending(d => d.FECHADISTRIBUCION).ToList();

            }
        }

        public static void CargarCerradosALista(ref DropDownList lista,
            int idEstablecimientoDistribucion, int anio)
        {
            lista.DataSource = ObtenerTodosCerrados(idEstablecimientoDistribucion, anio)
                .ToDictionary(d => d.IDDISTRIBUCION, d => d.IDDISTRIBUCION +" - "+ d.DESCRIPCION);
            lista.DataTextField = "Value";
            lista.DataValueField = "Key";
            lista.DataBind();
        }

        public static void CargarCerradosTodosALista(ref DropDownList lista, int idEstablecimiento, int anio)
        {
            anio --;
            using (var db = new SinabEntities())
            {
                var res = db.SAB_EST_DISTRIBUCION
                    .Where(
                        d =>
                            d.IDESTABLECIMIENTO == idEstablecimiento && d.FECHADISTRIBUCION.Year >= anio &&
                            (d.ESTADO == 2 || d.ESTADO == 3))
                    .Distinct().OrderByDescending(d => d.FECHADISTRIBUCION)
                    .Select(d => d).ToDictionary(d => d.IDDISTRIBUCION, d => d.IDDISTRIBUCION + " - " + d.DESCRIPCION);


                lista.DataSource = res;
                lista.DataTextField = "Value";
                lista.DataValueField = "Key";
                lista.DataBind();

            }
        }

        //public static List<SAB_EST_DISTRIBUCION> ObtenerReporte(int idDistribucion, int idEstablecimiento,
        //    int idEstablecimientoDistribucion, int idProducto)
        //{
        //    using (var db = new SinabEntities())
        //    {
        //        var res = from d in db.SAB_EST_DISTRIBUCION
        //                  where d.IDDISTRIBUCION == idDistribucion
        //                  && d.ides
        //    }
        //}
    }
}

File: Projects/ABASTECIMIENTOS/ABASTECIMIENTOS_ENTIDADES/Entidades/DISTRIBUCION.vb
Public Class DISTRIBUCION
    Inherits entidadBase

    'Variables
    Private _IDESTABLECIMIENTO As Integer
    Private _FECHADISTRIBUCION As Date
    Private _IDDISTRIBUCION As Integer
    Private _IDSUMINISTRO As Integer
    Private _IDALMACEN As Integer
    Private _NOMBREALMACEN As String
    Private _DESCRIPCION As String
    Private _FECHACORTE As Date
    Private _MESESCPM As Integer
    Private _MESESCOBERTURA As Integer
    Private _MESESADMINCPM As Integer
    Private _MESESSEGURIDADCPM As Integer
    Private _ESTADO As Integer
    Private _IDTIPOESTABLECIMIENTO As Integer
    Private _SUMINISTRO As String
    Private _AUUSUARIOCREACION As String
    Private _AUFECHACREACION As DateTime
    Private _AUUSUARIOMODIFICACION As String
    Private _AUFECHAMODIFICACION As DateTime
    Private _IDPROGRAMA As Integer

    Public Property IDESTABLECIMIENTO() As Integer
        Get
            Return _IDESTABLECIMIENTO
        End Get
        Set(ByVal Value As Integer)
            _IDESTABLECIMIENTO = Value
        End Set
    End Property

    Public Property FECHADISTRIBUCION() As Date
        Get
            Return _FECHADISTRIBUCION
        End Get
        Set(ByVal Value As Date)
            _FECHADISTRIBUCION = Value
        End Set
    End Property

    Public Property IDDISTRIBUCION() As Integer
        Get
            Return _IDDISTRIBUCION
        End Get
        Set(ByVal Value As Integer)
            _IDDISTRIBUCION = Value
        End Set
    End Property

    Public Property IDSUMINISTRO() As Integer
        Get
            Return _IDSUMINISTRO
        End Get
        Set(ByVal Value As Integer)
            _IDSUMINISTRO = Value
        End Set
    End Property

    Public Property IDALMACEN() As Integer
        Get
            Return _IDALMACEN
        End Get
        Set(ByVal Value As Integer)
            _IDALMACEN = Value
        End Set
    End Property

    Public Property NOMBREALMACEN() As String
        Get
            Return _NOMBREALMACEN
        End Get
        Set(ByVal Value As String)
            _NOMBREALMACEN = Value
        End Set
    End Property

    Public Property DESCRIPCION() As String
        Get
            Return _DESCRIPCION
        End Get
        Set(ByVal Value As String)
            _DESCRIPCION = Value
        End Set
    End Property

    Public Property FECHACORTE() As Date
        Get
            Return _FECHACORTE
        End Get
        Set(ByVal Value As Date)
            _FECHACORTE = Value
        End Set
    End Property

    Public Property MESESCPM() As Integer
        Get
            Return _MESESCPM
        End Get
        Set(ByVal Value As Integer)
            _MESESCPM = Value
        End Set
    End Property

    Public Property MESESCOBERTURA() As Integer
        Get
            Return _MESESCOBERTURA
        End Get
        Set(ByVal Value As Integer)
            _MESESCOBERTURA = Value
        End Set
    End Property

    Public Property MESESADMINCPM() As Integer
        Get
            Return _MESESADMINCPM
        End Get
        Set(ByVal Value As Integer)
            _MESESADMINCPM = Value
        End Set
    End Property

    Public Property MESESSEGURIDADCPM() As Integer
        Get
            Return _MESESSEGURIDADCPM
        End Get
        Set(ByVal Value As Integer)
            _MESESSEGURIDADCPM = Value
        End Set
    End Property

    Public Property ESTADO() As Integer
        Get
            Return _ESTADO
        End Get
        Set(ByVal Value As Integer)
            _ESTADO = Value
        End Set
    End Property

    Public Property IDTIPOESTABLECIMIENTO() As Integer
        Get
            Return _IDTIPOESTABLECIMIENTO
        End Get
        Set(ByVal Value As Integer)
            _IDTIPOESTABLECIMIENTO = Value
        End Set
    End Property

    Public Property SUMINISTRO() As String
        Get
            Return _SUMINISTRO
        End Get
        Set(ByVal Value As String)
            _SUMINISTRO = Value
        End Set
    End Property

    Public Property AUUSUARIOCREACION() As String
        Get
            Return _AUUSUARIOCREACION
        End Get
        Set(ByVal Value As String)
            _AUUSUARIOCREACION = Value
        End Set
    End Property

    Public Property AUFECHACREACION() As DateTime
        Get
            Return _AUFECHACREACION
        End Get
        Set(ByVal Value As DateTime)
            _AUFECHACREACION = Value
        End Set
    End Property

    Public Property AUUSUARIOMODIFICACION() As String
        Get
            Return _AUUSUARIOMODIFICACION
        End Get
        Set(ByVal Value As String)
            _AUUSUARIOMODIFICACION = Value
        End Set
    End Property

    Public Property AUFECHAMODIFICACION() As DateTime
        Get
            Return _AUFECHAMODIFICACION
        End Get
        Set(ByVal Value As DateTime)
            _AUFECHAMODIFICACION = Value
        End Set
    End Property
    Public Property IDPROGRAMA() As Integer
        Get
            Return _IDPROGRAMA
        End Get
        Set(ByVal Value As Integer)
            _IDPROGRAMA = Value
        End Set
    End Property
End Class
